src/analyses/ai.cpp

```diff
@@ -327,118 +317,174 @@ bool ai_baset::visit_edge(
   return false;
 }
 
-bool ai_baset::do_function_call(
+bool ai_baset::visit_edge_function_call(
   const irep_idt &calling_function_id,
   locationt l_call,
   locationt l_return,
+  const irep_idt &,
+  working_sett &working_set,
+  const goto_programt &,
+  const goto_functionst &,
+  const namespacet &ns)
+{
+  // The default implementation is not interprocedural
+  // so the effects of the call are approximated but nothing else
+  return visit_edge(
+    calling_function_id,
+    l_call,
+    calling_function_id,
+    l_return,
+    ns,
+    working_set);
+}
+
+bool ai_baset::visit_function_call(
+  const irep_idt &calling_function_id,
+  locationt l_call,
+  working_sett &working_set,
+  const goto_programt &caller,
   const goto_functionst &goto_functions,
-  const goto_functionst::function_mapt::const_iterator f_it,
-  const exprt::operandst &,
   const namespacet &ns)
 {
   PRECONDITION(l_call->is_function_call());
 
-  const goto_functionst::goto_functiont &goto_function=
-    f_it->second;
-
-  // This branch is for legacy support only and is deprecated
+  locationt l_return = std::next(l_call);
 
-  if(!goto_function.body_available())
+  // operator() allows analysis of a single goto_program independently
+  // it generates a synthetic goto_functions object for this
+  if(!goto_functions.function_map.empty())
   {
-    working_sett working_set; // Redundant; visit will add l_return
+    const code_function_callt &code = to_code_function_call(l_call->code);
+    const exprt &callee_expression = code.function();
 
-    // If we don't have a body, we just do an edge call -> return
-    return visit_edge(
-      calling_function_id,
-      l_call,
-      calling_function_id,
-      l_return,
-      ns,
-      working_set);
+    if(callee_expression.id() == ID_symbol)
+    {
+      const irep_idt &callee_function_id =
+        to_symbol_expr(callee_expression).get_identifier();
+
+      goto_functionst::function_mapt::const_iterator it =
+        goto_functions.function_map.find(callee_function_id);
+
+      DATA_INVARIANT(
+        it != goto_functions.function_map.end(),
+        "Function " + id2string(callee_function_id) + "not in function map");
+
+      const goto_functionst::goto_functiont &callee_fun = it->second;
+
+      if(callee_fun.body_available())
+      {
+        return visit_edge_function_call(
+          calling_function_id,
+          l_call,
+          l_return,
+          callee_function_id,
+          working_set,
+          callee_fun.body,
+          goto_functions,
+          ns);
+      }
+      else
+      {
+        // Fall through to the default, body not available, case
+      }
+    }
+    else
+    {
+      // Function pointers are not currently supported and must be removed
```

Comment:
This is a strange structure of code: do you plan on implementing this else-branch later?

Reply:
Yes.  I wanted to keep this patch set functionality neutral but with the work you and your team have done, this is *very* much on the cards.